docs/develop/optimizations.rst

```diff
@@ -0,0 +1,52 @@
+.. _optimizations:
+
+Optimizations
+=============
+
+MicroPython uses several optimizations to save RAM but also ensure the efficient
+execution of programs. This chapter discusses some of these optimizations.
+
+.. note::
+   :ref:`qstr` and :ref:`maps` details other optimizations on strings and 
+   dictionaries.
+
+
+Frozen bytecode
+----------------
+
+For efficiency, it is recommended that we avoid RAM or heap storage
```

Comment:
This is a topic that causes a lot of confusion in the forums. Can I propose a slight re-wording of this just to address one common source of confusion:

```
When MicroPython loads Python code from the filesystem, it first has to parse the file into a temporary in-memory representation, and then generate bytecode for execution, both of which are stored in the heap (in RAM). This can lead to significant amounts of memory being used.

The <link to glossary for mpy-cross> MicroPython cross compiler can be used to generate a ``.mpy`` file, containing the pre-compiled bytecode for a Python module. This will still be loaded into RAM, but it avoids the additional overhead of the parsing stage.

As a further optimisation, the pre-compiled bytecode from a ``.mpy`` file can be "frozen" into the firmware image as part of the main firmware compilation process, which means that the bytecode will be executed from ROM. This can lead to a significant memory saving, and reduce heap fragmentation.
```